presto-hive/src/test/java/com/facebook/presto/hive/TestBackgroundHiveSplitLoader.java

```diff
@@ -243,6 +262,65 @@ private static BackgroundHiveSplitLoader backgroundHiveSplitLoader(
                 false);
     }
 
+    private static BackgroundHiveSplitLoader backgroundHiveSplitLoaderOfflinePartitions()
+            throws Exception
+    {
+        ConnectorSession connectorSession = new TestingConnectorSession(
+                new HiveSessionProperties(new HiveClientConfig().setMaxSplitSize(new DataSize(1.0, GIGABYTE))).getSessionProperties());
+
+        return new BackgroundHiveSplitLoader(
+                SIMPLE_TABLE,
+                createPartitionMetadataWithOfflinePartitions(),
+                TupleDomain.all(),
+                createBucketSplitInfo(Optional.empty(), ImmutableList.of()),
+                connectorSession,
+                new TestingHdfsEnvironment(),
+                new NamenodeStats(),
+                new TestingDirectoryLister(TEST_FILES),
+                directExecutor(),
+                2,
+                false);
+    }
+
+    private static Iterable<HivePartitionMetadata> createPartitionMetadataWithOfflinePartitions()
+            throws RuntimeException
+    {
+        return new Iterable<HivePartitionMetadata>()
```

Comment:
I would guess you can use a lambda here

Reply:
Or move `hivePartitionMetadatas` outside of the `Iterable` declaration and convert it to:

a) lambda
```
return () -> hivePartitionMetadatas.stream()
  .map(......) // or .peek(...)
  .iterator()
```

b) `FluentIterable`:
```
return FluentIterable.from(hivePartitionMetadatas)
  .transform(...);
```

Reply:
Great idea with using `FluentIterable`. However in guava docs you may find for it: `A discouraged (but not deprecated) precursor to Java's superior Stream library.`.

You can achieve something similar with:
```
return hivePartitionMetadatas.stream()
     .map(...)
     ::iterator;
```

Reply:
As @findepi noticed returning `Stream::iterator` is not a good idea, as that method can be called only once.

Reply:
>A stream is single-use; it becomes invalid as soon as any "terminal operation" such as findFirst() or iterator() is invoked. (Even though Stream contains all the right method signatures to implement Iterable, it does not actually do so, to avoid implying repeat-iterability.) FluentIterable, on the other hand, is multiple-use, and does implement Iterable.
Streams offer many features not found here, including min/max, distinct, reduce, sorted, the very powerful collect, and built-in support for parallelizing stream

Reply:
Updated to use a lambda per @dain 's suggestion. For this test in particular, the method: `invokeNoMoreSplitsIfNecessary` needs to be called. The iterator is crafted (and has been simplified after feedback from @haozhun) to make sure it can touch the problematic code.